cmd/pulumi-test-language/l2destroy_test.go

```diff
@@ -0,0 +1,280 @@
+	}
+
+	return &pulumirpc.GetProgramDependenciesResponse{
+		Dependencies: []*pulumirpc.DependencyInfo{
+			{
+				Name:    "pulumi_pulumi",
+				Version: "1.0.1",
+			},
+			{
+				Name:    "pulumi_simple",
+				Version: "2.0.0",
+			},
+		},
+	}, nil
+}
+
+func (h *L2DestroyLanguageHost) InstallDependencies(
+	req *pulumirpc.InstallDependenciesRequest, server pulumirpc.LanguageRuntime_InstallDependenciesServer,
+) error {
+	if req.Info.RootDirectory != filepath.Join(h.tempDir, "projects", "l2-destroy", strconv.Itoa(h.currentRun)) {
+		return fmt.Errorf("unexpected root directory to install dependencies %s", req.Info.RootDirectory)
+	}
+	if req.Info.ProgramDirectory != req.Info.RootDirectory {
+		return fmt.Errorf("unexpected program directory to install dependencies %s", req.Info.ProgramDirectory)
+	}
+	if req.Info.EntryPoint != "." {
+		return fmt.Errorf("unexpected entry point to install dependencies %s", req.Info.EntryPoint)
+	}
+	return nil
+}
+
+func (h *L2DestroyLanguageHost) Run(
+	ctx context.Context, req *pulumirpc.RunRequest,
+) (*pulumirpc.RunResponse, error) {
+	if req.Info.RootDirectory != filepath.Join(h.tempDir, "projects", "l2-destroy", strconv.Itoa(h.currentRun)) {
+		return nil, fmt.Errorf("unexpected root directory to run %s", req.Info.RootDirectory)
+	}
+	if req.Info.ProgramDirectory != req.Info.RootDirectory {
+		return nil, fmt.Errorf("unexpected program directory to run %s", req.Info.ProgramDirectory)
+	}
+	if req.Info.EntryPoint != "." {
+		return nil, fmt.Errorf("unexpected entry point to run %s", req.Info.EntryPoint)
+	}
+
+	conn, err := grpc.Dial(
+		req.MonitorAddress,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		rpcutil.GrpcChannelOptions(),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("could not connect to resource monitor: %w", err)
+	}
+	defer conn.Close()
+
+	monitor := pulumirpc.NewResourceMonitorClient(conn)
+
+	_, err = monitor.RegisterResource(ctx, &pulumirpc.RegisterResourceRequest{
+		Type: string(resource.RootStackType),
+		Name: req.Stack,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("could not register stack: %w", err)
+	}
+
+	_, err = monitor.RegisterResource(ctx, &pulumirpc.RegisterResourceRequest{
+		Type:   "simple:index:Resource",
+		Custom: true,
+		Name:   "aresource",
+		Object: &structpb.Struct{
+			Fields: map[string]*structpb.Value{
+				"value": structpb.NewBoolValue(true),
+			},
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("could not register resource: %w", err)
+	}
+
+	if h.currentRun == 0 {
+		_, err = monitor.RegisterResource(ctx, &pulumirpc.RegisterResourceRequest{
+			Type:   "simple:index:Resource",
+			Custom: true,
+			Name:   "other",
+			Object: &structpb.Struct{
+				Fields: map[string]*structpb.Value{
+					"value": structpb.NewBoolValue(true),
+				},
+			},
+		})
+		if err != nil {
+			return nil, fmt.Errorf("could not register resource: %w", err)
+		}
+	}
+
+	h.currentRun++
```

Comment:
Doesn't Run get called twice per update? Once for the engine to preview then again for the actual up? We should definitely assert what we expect currentRun to be here. I think it'll be 0,1,2,3 and the if statement above should be for currentRun == 0 or == 1.

Reply:
No, we don't currently don't run previews in the conformance tests as far as I can see.  We're setting `SkipPreview` to true [here](https://github.com/pulumi/pulumi/blob/master/cmd/pulumi-test-language/interface.go#L880), and then only run [`stack.Update`](https://github.com/pulumi/pulumi/blob/master/cmd/pulumi-test-language/interface.go#L897), never `stack.Preview`.

Reply:
Ok we should still assert that currentRun is only ever 0 or 1 then.

Reply:
Probably at the very start of the function